### DISCUSSION TITLE
How to install a specific workload version?

### DISCUSSION THREAD

**Discussion by nebula2** [dotnet/maui#15390] [Q&A]
Hey :)

we are experiencing heavy issues with 7.0.86 (Grids have sadly become unusable). Tried .net 8 but along with that, other issues came. 
I want to downgrade to 7.0.81 - but I do not know how to do that.

I tried this:

1. 
```pwsh
dotnet workload uninstall maui-android maui-ios maui-windows maui-maccatalyst
```

2. 
```pwsh
dotnet workload install maui-android maui-ios maui-windows maui-maccatalyst `
--sdk-version=7.0.302 `
--no-cache `
--verbosity detailed `
--from-rollback-file https://maui.blob.core.windows.net/metadata/rollbacks/7.0.81.json `
--source https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet7/nuget/v3/index.json `
--source https://api.nuget.org/v3/index.json
```

If I now do `dotnet workload list` I receive the following output:

```text
maui-android               7.0.86/7.0.100         SDK 7.0.300, VS 17.6.33723.286
maui-ios                   7.0.86/7.0.100         SDK 7.0.300, VS 17.6.33723.286
maui-maccatalyst           7.0.86/7.0.100         SDK 7.0.300, VS 17.6.33723.286
maui-windows               7.0.86/7.0.100         SDK 7.0.300, VS 17.6.33723.286
maccatalyst                16.4.7060/7.0.100      VS 17.6.33723.286
ios                        16.4.7060/7.0.100      VS 17.6.33723.286
android                    33.0.46/7.0.100        VS 17.6.33723.286

Use `dotnet workload search` to find additional workloads to install.

Updates are available for the following workload(s): maui-android maui-ios maui-maccatalyst. Run `dotnet workload update` to get the latest.
```

It says `7.0.86/7.0.100`. I was expecting `7.0.81/something`. Can someone please guide me on how to get this going?

---
EDIT:

I just checked #12998
and added mauiversion even though not really supported - just for the lulz.

I also checked installed sdk's and well - as expected no 7.0.81 there.
```
❯ dir 'C:\Program Files\dotnet\packs\Microsoft.Maui.Sdk\'

    Directory: C:\Program Files\dotnet\packs\Microsoft.Maui.Sdk

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d----          10.05.2023    11:37                6.0.553
d----          10.05.2023    11:37                7.0.86
```

**Comment by nebula2** [comment]
Is this how I can interpret the absence of responses?

`No one has considered the eventuality that something unusable will be released.`

**Reply by FM1973** [reply]
We are using #12998 as well and it works.
I know it´s not officially supported, but currently the only way to use an older version like i.e. 7.0.59.
As far as I understand, your app gets compiled against the current version, but uses the <MauiVersion> at runtime.
They will support nuget packages later. So anyone can pin to a certain nuget package. But this will not be available before .net 8, I think.

**Reply by nebula2** [reply]
> We are using #12998 as well and it works.
> I know it´s not officially supported, but currently the only way to use an older version like i.e. 7.0.59.
> As far as I understand, your app gets compiled against the current version, but uses the <MauiVersion> at runtime.
> They will support nuget packages later. So anyone can pin to a certain nuget package. But this will not be available before .net 8, I think.

thanks for your answer. I am however struggling to even install an older workload version on my machines :/

Do you know how to do that?

**Comment by arkfen** [comment]
it actually worked for me, so thank you!

I used

sudo dotnet workload install maui-maccatalyst --sdk-version=7.0.302 --from-rollback-file https://maui.blob.core.windows.net/metadata/rollbacks/7.0.81.json            

result:
Installed Workload Id      Manifest Version      Installation Source
--------------------------------------------------------------------
maui-maccatalyst           7.0.81/7.0.100        SDK 7.0.300   

please note that before doing that I removed ALL dotnet 100%, then installed Dotnet 7.0.5 (SDK 7.0.302) and then the command above... I guess maybe your uninstall command did not do a good job or somethings else... I hope you will be able to figure it out.

At the moment these tricks can be needed for lots of people who use older Macs, because XCode 14.3 asks for latest MacOS which cant be installed on macs below 2017 (pro 2018) macbooks... so in order to work with latest maui workloads you need to chose "Link SDK only" during debuging as the work around ... or just roll back to old ones as I try to do now

**Reply by beeradmoore** [reply]
This solution no longer works as that url gives the error

> Public access is not permitted on this storage account.

**Comment by bassuny3003** [comment]
thanks

new in road of maui

**Comment by AceCoderLaura** [comment]
In dotnet SDK 8.0.400 and higher you can now downgrade to a previous workload version using `dotnet workload update --version 9.0.305`, hope this helps anyone who stumbles upon this discussion.

**Reply by Ghostbird** [reply]
How does this help me install a specific MAUI workload version?

**Reply by AceCoderLaura** [reply]
@Ghostbird MAUI workloads will be downgraded to the specified version when you downgrade your workloads. As far as I can tell, you can't downgrade a _selection_ of workloads without causing compatibility issues, you have to downgrade all your workloads at once.

**Reply by Ghostbird** [reply]
Kind of annoying, I understand this, and I can see why it's like that. Alas, it doesn't provide me with a good solution.

My use case is that I build docker images that can be used as an Android build environment, for a given MAUI version. I'd much rather have an option to request the installation of a specific MAUI version and have it fail when dependency resolution fails. For now I simply build the image for the most recent MAUI version, and I cannot create images for older versions. So far this has not resulted in problems.
